File: src/models/constraints/overall_var_constraints.jl
"""
Modeling module
Declare overall variable constraints (as sum of unit variables).

Pengfei Cheng

1. constraints for overall variables (8)
        load factor             fuel rate               CO2 emission rate
        steam turbine power     auxiliary power
"""

function add_overall_var_constraints(m)
    x_load = m[:x_load]
    x_load_D = m[:x_load_D]
    x_fuel = m[:x_fuel]
    x_fuel_D = m[:x_fuel_D]
    x_CO2_flue = m[:x_CO2_flue]
    x_CO2_D_flue = m[:x_CO2_D_flue]
    x_power_ST = m[:x_power_ST]
    x_power_D_ST = m[:x_power_D_ST]
    x_power_aux = m[:x_power_aux]
    x_power_D_aux = m[:x_power_D_aux]

    # 1. CONSTRAINTS FOR OVERALL VARIABLES

    # load factor
    @constraint(
        m ,eq_total_load_factor[i = set_hour_0],
        x_load[i]
        ==
        sum(x_load_D[i, mode] for mode in set_mode)
    )

    # fuel consumption
    @constraint(
        m, eq_total_fuel_rate[i = set_hour_0], 
        x_fuel[i]
        ==
        sum(x_fuel_D[i, mode] for mode in set_mode)
    )

    # CO2 emission rate
    @constraint(
        m, eq_total_CO2_emission_rate[i = set_hour_0], 
        x_CO2_flue[i]
        ==
        sum(x_CO2_D_flue[i, mode] for mode in set_mode)
    )

    # steam turbine power
    @constraint(
        m, eq_total_power_ST[i = set_hour_0], 
        x_power_ST[i]
        ==
        sum(x_power_D_ST[i, mode] for mode in set_mode)
    )

    # auxiliary power
    @constraint(
        m, eq_total_power_aux[i = set_hour_0], 
        x_power_aux[i]
        ==
        sum(x_power_D_aux[i, mode] for mode in set_mode)
    )

end

File: src/pyomo_model/constraints/overall_var_constraints.py
"""
Modeling module
Declare overall variable constraints (as sum of unit variables).

Pengfei Cheng

1. constraints for overall variables (8)
        load factor             fuel rate               CO2 emission rate
        HP steam turbine power  IP steam turbine power  DAC base steam
        allocable steam
"""

from pyomo.environ import Constraint
from ..params import *


def add_overall_var_constraints(m, set_hour_0, set_scenario):

    # 1. CONSTRAINTS FOR OVERALL VARIABLES

    # load factor
    def eq_total_load_factor(m, i, s):
        return m.x_load[i, s] == sum(m.x_load_D[i, mode, s] for mode in set_mode)
    m.eq_total_load_factor = Constraint(set_hour_0, set_scenario, rule=eq_total_load_factor)

    # fuel consumption
    def eq_total_fuel_rate(m, i, s):
        return m.x_fuel[i, s] == sum(m.x_fuel_D[i, mode, s] for mode in set_mode)
    m.eq_total_fuel_rate = Constraint(set_hour_0, set_scenario, rule=eq_total_fuel_rate)

    # CO2 emission rate
    def eq_total_CO2_emission_rate(m, i, s):
        return m.x_CO2_flue[i, s] == sum(m.x_CO2_D_flue[i, mode, s] for mode in set_mode)
    m.eq_total_CO2_emission_rate = Constraint(set_hour_0, set_scenario, rule=eq_total_CO2_emission_rate)

    # HP steam turbine power
    def eq_total_power_HP(m, i, s):
        return m.x_power_HP[i, s] == sum(m.x_power_D_HP[i, mode, s] for mode in set_mode)
    m.eq_total_power_HP = Constraint(set_hour_0, set_scenario, rule=eq_total_power_HP)

    # IP steam turbine power
    def eq_total_power_IP(m, i, s):
        return m.x_power_IP[i, s] == sum(m.x_power_D_IP[i, mode, s] for mode in set_mode)
    m.eq_total_power_IP = Constraint(set_hour_0, set_scenario, rule=eq_total_power_IP)

    # DAC base steam
    def eq_total_DAC_base_duty(m, i, s):
        return m.x_steam_DAC_base[i, s] == sum(m.x_steam_D_DAC_base[i, mode, s] for mode in set_mode)
    m.eq_total_DAC_base_duty = Constraint(set_hour_0, set_scenario, rule=eq_total_DAC_base_duty)

    # allocable steam
    def eq_total_steam_allocable(m, i, s):
        return m.x_steam_allocable[i, s] == sum(m.x_steam_D_allocable[i, mode, s] for mode in set_mode)
    m.eq_total_steam_allocable = Constraint(set_hour_0, set_scenario, rule=eq_total_steam_allocable)

    # auxiliary power
    def eq_total_power_aux(m, i, s):
        return m.x_power_aux[i, s] == sum(m.x_power_D_aux[i, mode, s] for mode in set_mode)
    m.eq_total_power_aux = Constraint(set_hour_0, set_scenario, rule=eq_total_power_aux)

    return